src/primitives/hrpstring.rs

```diff
@@ -0,0 +1,460 @@
+// Written by Clark Moody and the rust-bitcoin developers.
+// SPDX-License-Identifier: MIT
+
+//! String encoding/decoding of HRP strings as specified by [BIP-173] (bech32) and [BIP-350] (bech32m).
+//!
+//! HRP string format: `<hrp> 1 <witness_version> <payload> <checksum>`
+//!
+//! * **hrp**: Human Readable Part.
+//! * **witness_version**: 0 ('q) for bech32 1 ('p') for bech32m.
+//! * **payload**: GF(32) bech32 encoded characters.
+//! * **checksum**: BCH code checksum.
+//!
+//! [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+//! [BIP-350]: <https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki>
+
+use core::convert::{Infallible, TryFrom};
+use core::iter::FusedIterator;
+use core::marker::PhantomData;
+use core::{fmt, iter, slice, str};
+
+use internals::write_err;
+
+use crate::network::Network;
+use crate::primitives::checksum::{self, Checksum};
+use crate::primitives::gf32::Fe32;
+use crate::primitives::iter::{Fe32IterExt, FeToByteIter};
+use crate::witness_version::{self, WitnessVersion};
+
+/// Maximum length of the human-readable part, as defined by BIP-173.
+const MAX_HRP_LEN: usize = 83;
+
+/// An HRP string that has been parsed from an ordinary checksummed string.
+///
+/// Parsing as an HRP string does not validate the checksum in any way.
+#[derive(Debug)]
+pub struct Parsed<'s> {
```

Comment:
The names here don't make a ton of sense. The module is called `hrpstring`, but the primary data structure is `Parsed` and contains not only HRP data but also witness version and the data part.

Reply:
The data structure is an HRP string which has been parsed into an HRP, witness version, and data part.

We should change this if it's not understandable but it seemed pretty straightforward to me.